server/sonar-server/src/main/java/org/sonar/server/settings/ws/ValuesAction.java

```diff
@@ -0,0 +1,213 @@
+      this.dbSession = dbSession;
+      this.valuesWsBuilder = valuesWsBuilder;
+      this.keys = keys;
+      this.component = component;
+    }
+
+    void build() {
+      processDefinitions();
+      processProperties(settingsLoader.loadGlobalSettings(dbSession, keys));
+      if (component != null) {
+        processProperties(settingsLoader.loadComponentSettings(dbSession, keys, component).values());
+      }
+      settingsBuilderByKey.values().forEach(Settings.Setting.Builder::build);
+    }
+
+    private void processDefinitions() {
+      propertyDefinitions.getAll().stream()
+        .filter(definition -> keys.contains(definition.key()))
+        .filter(defaultProperty -> !isNullOrEmpty(defaultProperty.defaultValue()))
+        .map(Setting::new)
+        .forEach(this::processDefaultValue);
+    }
+
+    private void processDefaultValue(Setting setting) {
+      Settings.Setting.Builder settingBuilder = valuesWsBuilder.addSettingsBuilder()
+        .setKey(setting.getKey())
+        .setDefault(true);
+      setValue(setting, settingBuilder);
+      settingsBuilderByKey.put(setting.getKey(), settingBuilder);
+    }
+
+    private void processProperties(Collection<Setting> properties) {
+      properties.forEach(setting -> {
+        String key = setting.getKey();
+        Settings.Setting.Builder valueBuilder = getOrCreateValueBuilder(key);
+        valueBuilder.setInherited(component != null && !ObjectUtils.equals(setting.getComponentId(), component.getId()));
+        valueBuilder.setDefault(false);
+        setValue(setting, valueBuilder);
+      });
+    }
+
+    private Settings.Setting.Builder getOrCreateValueBuilder(String propertyKey) {
+      Settings.Setting.Builder valueBuilder = settingsBuilderByKey.get(propertyKey);
+      if (valueBuilder == null) {
+        valueBuilder = valuesWsBuilder.addSettingsBuilder().setKey(propertyKey);
+        settingsBuilderByKey.put(propertyKey, valueBuilder);
+      }
+      return valueBuilder;
+    }
+
+    private void setValue(Setting setting, Settings.Setting.Builder valueBuilder) {
+      PropertyDefinition definition = setting.getDefinition();
+      String value = setting.getValue();
+      if (definition == null) {
+        valueBuilder.setValue(value);
+        return;
+      }
+      if (definition.type().equals(PROPERTY_SET)) {
+        Settings.FieldsValues.Builder builder = Settings.FieldsValues.newBuilder();
+        for (Map<String, String> propertySetMap : getPropertySetValuesAsMap(setting)) {
+          builder.addFieldsValuesBuilder().putAllValue(propertySetMap);
+        }
+        valueBuilder.setFieldsValues(builder);
+      } else if (definition.multiValues()) {
+        valueBuilder.setValues(Settings.Values.newBuilder().addAllValues(COMMA_SPLITTER.split(value)));
+      } else {
+        valueBuilder.setValue(value);
+      }
+    }
+
+    private List<Map<String, String>> getPropertySetValuesAsMap(Setting setting) {
```

Comment:
extract this logic to `Setting` or `SettingsLoader`. I prefer Setting, so the getter return a list of map